### DISCUSSION TITLE
Running extra CKAN containers easily

### DISCUSSION THREAD

**Discussion by frafra** [ckan/ckan#6489] [Ideas]
Hi!
I use ckanext-harvest, but I always need to mess around with `ckan-entrypoint.sh`, as I do not want to have the database being initialized multiple times concurrently, which causes errors (duplicate pk and such - probably some transaction are missing).

What about skipping db init for such extra containers?

Proposed PR: https://github.com/ckan/ckan/pull/6490

**Comment by ricardogsilva** [comment]
IMO removing the DB initialization completely from `ckan-entrypoint.sh` would be even better.

For what its worth, on a project we are working on, we've implemented a docker entrypoint script in Python rather than bash that does not call `ckan db init`

https://github.com/kartoza/ckanext-dalrrd-emc-dcpr/blob/main/ckanext/dalrrd_emc_dcpr/cli/docker_entrypoint.py

This means we need to explicitly initialize the DB after standing up the ckan stack - but since this is only done once it doesn't seem like a big drawback

**Reply by frafra** [reply]
CKAN container should be able to start up from scratch without any other action in my opinion. Using a different command to start it the first time, and then changing again the compose file to run as usual doesn't seem a good method to me, if that is what you are suggesting.
We could have a Python entrypoint, but I am not sure about the benefits. Avoiding to export env variables maybe?

**Reply by ricardogsilva** [reply]
> CKAN container should be able to start up from scratch without any other action in my opinion

In our case, it does start up from scratch - it runs the gunicorn application server. Granted, CKAN's web page and API are not really usable until you initialize the DB. But the container itself is running, which means we can use the `ckan` CLI utility to run further commands, such as the DB initialization.

> Using a different command to start it the first time, and then changing again the compose file to run as usual doesn't seem a good method to me, if that is what you are suggesting.

That is not what I'm proposing at all. What I mean is that after the container has been stood up, we can now run something like:

```
docker exec -ti {container-name} poetry run ckan db init
docker exec -ti {container-name} poetry run ckan sysadmin add
# run any additional bootstrap commands
```

After standing up the stack for the first time it is likely that there will be a number of bootstrap commands that need to be ran (i.e. creating default organizations and users, etc.). The approach I mentioned just makes the DB initialization a first step in the bootstrap process. This assumes that the CKAN DB is being stored in some persistent volume and therefore survives between container spin-ups and downs.

> We could have a Python entrypoint, but I am not sure about the benefits. 

This is a matter of opinion. To me it seems logical that a Python project would use Python for this kind of task, as it lowers the cognitive load somewhat - I wouldn't introduce another language to the project unless absolutely necessary.

---

My thing with suggesting removing the DB initialization script from the docker entrypoint is that it seems that 99.9% of the time (_i.e._ after the stack has already been bootstrapped) it is not needed and that 0.1% where it is needed you will likely also want to run other commands in order to properly bootstrap the CKAN deployment, so you might as well initialize the DB manually. Furthermore, as you originally mentioned, having automatic DB initialization can become problematic when launching multiple CKAN containers at the same time.

**Reply by frafra** [reply]
Thanks for clarifying!
Sure, a different container for the initialization could be executed alongside and added to the compose file, but you need to wait for the database to go up, to avoid timing issues.
The main issue today is that the database initialization is not made in a single transaction nor can be executed in parallel. Otherwise, it would be fine to try to initialize the database each time.
